### PR TITLE
Add support for pruned models

### PR DESCRIPTION
**Pull request by MajorCarrot**
According to the [pytorch documentation on pruning](https://pytorch.org/tutorials/intermediate/pruning_tutorial.html),
the original parameter is replaced with one ending with `_orig` and
a new buffer ending with `_mask`. The mask contains 0s and 1s based on
which the correct parameters are chosen.

All instances of `param.nelements()` have been replaced by a variable
`cur_params` whose value is set based on whether it is a masked model or
not. To keep consistency with the rest of the code base, the `_orig` is
removed from the `name` variable right after the calculation of
`cur_params`.

### CODE REVIEW

**Comment by MajorCarrot** [comment]
I am not sure how to fix the mypy errors but I have tested this branch on pruned VGG networks and it seems to work. If you need any inputs from my side please ping me :)

File: torchinfo/layer_info.py [hunk 1]
```diff
@@ -121,9 +130,18 @@ def calculate_num_params(self) -> None:
...
+            # Masked models save the parameter with the name "_orig" added
+            # They have a buffer ending with "_mask" which has only 0s and 1s
+            if name[-4:] == "orig":
+                cur_params = int(torch.sum(rgetattr(self.module, name[:-4] + "mask")))
+                # Remove "_orig" for better readability and integration
+                name = name[:-5]
+            else:
+                cur_params = param.nelement()
+
+            self.num_params += cur_params
```
Comment (TylerYep) [inline torchinfo/layer_info.py:142]: Let's move this into a function so we don't need to copy paste the code

File: torchinfo/layer_info.py [hunk 2]
```diff
@@ -10,6 +11,14 @@
 ]
 
 
+# using wonder's beautiful simplification: https://stackoverflow.com/a/31174427/8890157
+def rgetattr(obj: object, attr: str, *args) -> object:
```
Comment (TylerYep) [inline torchinfo/layer_info.py:15]: Add `: Any` to *args type annotation

Comment (TylerYep) [reply torchinfo/layer_info.py:15]: Also I think obj is always of type `str`

**Comment by TylerYep** [comment]
@MajorCarrot Looks good so far. Could you add a test case with a pruned VGG network and the expected output?

Also left a couple comments on the changes

File: torchinfo/layer_info.py (hunk 2 above)
Comment (MajorCarrot) [reply torchinfo/layer_info.py:15]: `obj` would be the module itself whose submodule/attribute we are trying to extract. I am not sure if it would be a string.

**Comment by TylerYep** [comment]
FYI, you'll need to run `pytest --overwrite` in order to create the expected output file for the test case you added

**Comment by MajorCarrot** [comment]
I have fixed almost all the errors with mypy (by rewriting the rgetattr and changing its signature), but I am still not able to get the whole thing passed through mypy xD

If you have suggestions on how to fix this (hopefully final) error, we should be good to go!

File: torchinfo/layer_info.py (hunk 1 above)
Comment (MajorCarrot) [reply torchinfo/layer_info.py]: Resolving as change has been implemented

File: torchinfo/layer_info.py (hunk 2 above)
Comment (MajorCarrot) [reply torchinfo/layer_info.py]: I rewrote this whole thing and the issue is now fixed

**Comment by TylerYep** [comment]
You can add a `# type: ignore` on the last error. The issue is that this l2_unstructured function isn't type-annotated

File: tests/torchinfo_test.py [hunk 3]
```diff
@@ -162,6 +163,18 @@ def test_resnet152() -> None:
     summary(model, (1, 3, 224, 224), depth=3)
 
 
+def test_pruning() -> None:
+    model = SingleInputNet()
+    for module in model.modules():
+        if isinstance(module, (torch.nn.Conv2d, torch.nn.Linear)):
+            # mypy: allow-untyped-calls
```
Comment (TylerYep) [inline tests/torchinfo_test.py:170]: Remove this line, needs to be `# type ignore[allow-untyped-calls]` on the next line

File: torchinfo/layer_info.py [hunk 4]
```diff
@@ -12,6 +12,17 @@
 ]
 
 
+def rgetattr(obj: torch.nn.Module, attr: str) -> torch.Tensor:
+    """Get the tensor submodule called attr from obj."""
+    for i in attr.split("."):
```
Comment (TylerYep) [inline torchinfo/layer_info.py:17]: Let's use a dfferent variable since `i` is typically reserved for numbers

File: torchinfo/layer_info.py [hunk 5]
```diff
@@ -116,16 +127,31 @@ def get_layer_name(self, show_var_name: bool, show_depth: bool) -> str:
...
+    def __get_cur_params(self, name: str, param: torch.Tensor) -> tuple[int, str]:
+        """
+        Get count of number of params, accounting for mask
+        """
+        # Masked models save the parameter with the name "_orig" added
+        # They have a buffer ending with "_mask" which has only 0s and 1s
+        if name[-4:] == "orig":
+            # If a mask exists, the sum of 1s in mask is number of params
+            # Remove "_orig" for better readability and integration
+            return int(torch.sum(rgetattr(self.module, name[:-4] + "mask"))), name[:-5]
```
Comment (TylerYep) [inline torchinfo/layer_info.py:139]:
```suggestion
            return int(torch.sum(rgetattr(self.module, f"{name[:-4]}mask"))), name[:-5]
```

**Review by TylerYep — COMMENTED** [review]
Couple more changes

File: torchinfo/layer_info.py (hunk 4 above)
Comment (MajorCarrot) [reply torchinfo/layer_info.py:17]: Would `attr_i` be fine (to say some part of the variable `attr`)?

**Comment by MajorCarrot** [comment]
(I am really sorry for messing up the whole testing thing, so many runs :facepalm:)

I had removed the `0` in `0.5` because flake8 doesn't like the longer line and I can't shift the `# type: ignore` anywhere else

File: tests/torchinfo_test.py [hunk 6]
```diff
@@ -162,6 +163,17 @@ def test_resnet152() -> None:
     summary(model, (1, 3, 224, 224), depth=3)
 
 
+def test_pruning() -> None:
+    model = SingleInputNet()
+    for module in model.modules():
+        if isinstance(module, (torch.nn.Conv2d, torch.nn.Linear)):
+            prune.l1_unstructured(module, "weight", 0.5)  # type: ignore[no-untyped-call]
```
Comment (TylerYep) [inline tests/torchinfo_test.py:170]:
```suggestion
            prune.l1_unstructured(  # type: ignore[no-untyped-call]
                module, "weight", 0.5
            )
```

Comment (TylerYep) [reply tests/torchinfo_test.py:170]: Try this

Comment (MajorCarrot) [reply tests/torchinfo_test.py]: I think it worked! Thanks :)

**Comment by TylerYep** [comment]
Thanks for the contribution! Please feel free to continue finding improvements :)

**Comment by TylerYep** [comment]
This is now available in torchinfo v1.6.1
